pandas/core/dtypes/missing.py

```diff
@@ -506,14 +506,27 @@ def array_equals(left: ArrayLike, right: ArrayLike) -> bool:
         return array_equivalent(left, right, dtype_equal=True)
 
 
-def infer_fill_value(val):
+def infer_fill_value(val, length: int):
     """
     infer the fill value for the nan/NaT from the provided
     scalar/ndarray/list-like if we are a NaT, return the correct dtyped
     element to provide proper block construction
     """
     if not is_list_like(val):
         val = [val]
+
+    if is_extension_array_dtype(val):
+        # We cannot use dtype._na_value bc pd.NA/pd.NaT do not preserve dtype
+        if len(val) == length:
```

Comment:
i am not sure how this can every be not true.

if val is an array this is always true?
if its a scalar it is as well rigth?

Reply:
updated + green

Reply:
ill give this some thought in the AM.  this is a difficult function to describe.